pkg/jobs/wait.go

```diff
@@ -0,0 +1,229 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package jobs
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
+	"github.com/cockroachdb/cockroach/pkg/security"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlutil"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/errors"
+)
+
+// NotifyToAdoptJobs notifies the job adoption loop to start claimed jobs.
+func (r *Registry) NotifyToAdoptJobs(context.Context) {
+	select {
+	case r.adoptionCh <- resumeClaimedJobs:
+	default:
+	}
+}
+
+// NotifyToResume is used to notify the registry that it should attempt
+// to resume the specified jobs. The assumption is that these jobs were
+// created by this registry and thus are pre-claimed by it. This bypasses
+// the loop to discover jobs already claimed by this registry. If the jobs
+// turn out to not be claimed by this registry, it's not a problem.
+func (r *Registry) NotifyToResume(ctx context.Context, jobs ...jobspb.JobID) {
+	m := newJobIDSet(jobs...)
+	_ = r.stopper.RunAsyncTask(ctx, "resume-jobs", func(ctx context.Context) {
+		r.withSession(ctx, func(ctx context.Context, s sqlliveness.Session) {
+			r.filterAlreadyRunningAndCancelFromPreviousSessions(ctx, s, m)
+			r.resumeClaimedJobs(ctx, s, m)
+		})
+	})
+}
+
+// WaitForJobs waits for a given list of jobs to reach some sort
+// of terminal state.
+func (r *Registry) WaitForJobs(
+	ctx context.Context, ex sqlutil.InternalExecutor, jobs []jobspb.JobID,
+) error {
+	log.Infof(ctx, "waiting for %d %v queued jobs to complete", len(jobs), jobs)
+	done, cleanup := r.installWaitingSet(jobs...)
+	defer cleanup()
+	return r.waitForJobs(ctx, ex, jobs, done)
+}
+
+func (r *Registry) waitForJobs(
+	ctx context.Context, ex sqlutil.InternalExecutor, jobs []jobspb.JobID, done <-chan struct{},
```

Comment:
nit: (maybe silly one) ... I always associate done with something that this func closes when it's done.  But it appears that in this case done is a signal to wait to stop waiting.  perhaps rename to abortWait or some such?